File: solutions/python/linked-list/1/linked_list.py
""" Module docstring """
class Node:
    """ Class docstring """
    def __init__(self, value, succeeding=None, previous=None):
        """ function docstring """
        self.val = value
        self.nxt = succeeding
        self.prev = previous


class LinkedList:
    """ Class docstring """
    def __init__(self):
        """ function docstring """
        self.first = None
        self.last = None

    def push(self, value):
        """ function docstring """
        # insert at back
        cur = Node(value)
        cur.prev = self.last
        if self.last is not None:
            self.last.nxt = cur
        if self.first is None:
            self.first = cur
        self.last = cur

    def pop(self):
        """ function docstring """
        # remove from back
        if self.last is not None:
            curnode = self.last
            value = curnode.val
            if self.first == self.last:
                self.first = None
            self.last = curnode.prev
            if self.last is not None:
                self.last.nxt = None
            return value
        return None

    def shift(self):
        """ function docstring """
        # remove from front
        if self.first == self.last:
            self.last = None
        if self.first is not None:
            firstnode = self.first
            value = firstnode.val
            if firstnode.nxt is not None:
                self.first = firstnode.nxt
            return value

    def unshift(self, value):
        """ function docstring """
        # insert in front
        cur = Node(value)
        if self.first is not None:
            self.first.prev = cur
        cur.nxt = self.first
        self.first = cur
        if self.last is None:
            self.last = cur

    def printll(self):
        curnode = self.first
        while curnode is not None:
            print(curnode.val, end="|")
            curnode = curnode.nxt
        print()
lst = LinkedList()
lst.push(10)
lst.push(20)
print(lst.first.val,lst.first.nxt)
lst.printll()
print(lst.pop(), 20)
lst.printll()
lst.push(30)
lst.printll()
print(lst.shift(), 10)
lst.printll()
lst.unshift(40)
lst.printll()
lst.push(50)
lst.printll()
print(lst.shift(), 40)
lst.printll()
print(lst.pop(), 50)
lst.printll()
print(lst.shift(), 30)
lst.printll()

File: solutions/fortran/linked-list/1/linked_list.f90
module linked_list
  implicit none

  type :: list_t
    type(list_n), pointer :: first_node
  end type list_t

  type :: list_n
    integer :: val 
    type(list_n), pointer :: next_node
  end type list_n

contains

  function new() result(list)
    type(list_t), allocatable :: list
     
    allocate(list)
    nullify(list%first_node)
  end function new

  subroutine push(list, val)
    type(list_t), intent(inout) :: list
    integer, intent(in) :: val
    type(list_n) , target, allocatable :: list_node
    type(list_n) , pointer:: list_node_ptr
    type(list_n) , pointer:: tnode_ptr
    print *, "push"
    print *, "val=", val
    if (.not. associated(list%first_node)) then
        allocate(list_node)
        list_node%val = val
        nullify(list_node%next_node)
        list%first_node => list_node
        print *,"first node val=", list_node%val, ", associated(list%first_node)=",associated(list%first_node) 
    else      
       !ranverse the list to find the last node
       print *, "associated(list%first_node)=",associated(list%first_node) 
       tnode_ptr = list%first_node
       do while (associated(tnode_ptr%next_node))
         tnode_ptr = tnode_ptr%next_node
       end do

       allocate(list_node)
       tnode_ptr%next_node => list_node
       list_node%val = val
       nullify(list_node%next_node)
       print *, " rest of nodes, val=",list_node%val,",associated(list_node%next_node)=",associated(list_node%next_node) 
    end if
    print *, "end of push"

  end subroutine push

  function pop(list) result(val)
    type(list_t), intent(inout) :: list
    integer :: val
    type(list_n) , pointer:: tnode_ptr
    print *, "pop"
    print *, "associated(list%first_node)=",associated(list%first_node) 

    ! if list is empty just return
    if (.not. associated(list%first_node)) return
        
    ! if only a first node then delete it
    nullify(tnode_ptr)
    tnode_ptr => list%first_node
    print *, "val=",tnode_ptr%val,",associated(tnode_ptr%next_node)=",associated(tnode_ptr%next_node) 
    if (.not. associated(tnode_ptr%next_node)) then
            val = tnode_ptr%val
            print *, "2 val=", val
            deallocate(tnode_ptr)
            return
    end if
    print *,"shouldn't be here on first time"
    ! tranverse the list to find the last node
    print *, "2 val=",tnode_ptr%val, ", associated(tnode_ptr%next_node)=",associated(tnode_ptr%next_node) 
    do while (associated(tnode_ptr%next_node))
       tnode_ptr => tnode_ptr%next_node
       print *, "3 val=",tnode_ptr%val, ", associated(tnode_ptr%next_node)=",associated(tnode_ptr%next_node) 
    end do
    val = tnode_ptr%val
    deallocate(tnode_ptr)

  end function pop

  subroutine unshift(list, val)
    type(list_t), intent(inout) :: list
    integer, intent(in) :: val
    type(list_n), allocatable :: list_node

    ! add node to start of link list
    allocate(list_node)
    list_node%val = val
    list_node%next_node => list%first_node
    list%first_node = list_node

  end subroutine unshift

  function shift(list) result(val)
    type(list_t), intent(inout) :: list
    integer :: val
    type(list_n) , pointer:: tnode_ptr

    ! Remove node from start of list
    ! check if there is a node to remove
    if (.not. associated(list%first_node)) return
    
    tnode_ptr => list%first_node
    val = tnode_ptr%val
    list%first_node => tnode_ptr%next_node
    deallocate(tnode_ptr) 
    
  end function shift

  function length(list) result(n)
    type(list_t), intent(in) :: list
    integer :: n
    type(list_n) , pointer:: tnode_ptr
    
    n = 0

    ! tranverse the list to find the last node
    tnode_ptr => list%first_node
    do while (associated(tnode_ptr%next_node))
       n = n + 1
       tnode_ptr => tnode_ptr%next_node
    end do
  end function length

  subroutine delete(list, val)
    type(list_t), intent(inout) :: list
    integer :: val
    type(list_n) , pointer:: tnode_ptr1, tnode_ptr2
    ! If there arent any nodes then just return
    if (.not. associated(list%first_node)) return
    ! if the first node is the one to delete
    if (list%first_node%val == val) then
        deallocate(list%first_node)
        nullify(list%first_node)
        return
    endif    
    ! find the node with the value and delete it
    tnode_ptr1 => list%first_node
    tnode_ptr2 => tnode_ptr1%next_node
    do while (associated(tnode_ptr2%next_node))
       if (tnode_ptr2%val == val) then
           tnode_ptr1%next_node = tnode_ptr2%next_node    
           deallocate(tnode_ptr2)
           exit
       end if    
       tnode_ptr1 => tnode_ptr2
       tnode_ptr2 => tnode_ptr2%next_node
    end do

  end subroutine delete

  subroutine destroy(list)
    type(list_t), intent(inout) :: list
    type(list_n) , pointer:: tnode_ptr1, tnode_ptr2
    
    tnode_ptr1 => list%first_node
    do while (associated(tnode_ptr1%next_node))
           tnode_ptr2 => tnode_ptr1%next_node    
           deallocate(tnode_ptr1)
           tnode_ptr1 => tnode_ptr2
    end do
    !deallocate(list)

  end subroutine destroy

end module
